DBZ-4665 Extend Debezium Schema Generator

rk3rn3r opened the pull request:
* finish renaming of API/APIFormat occurences to use Schema instead
* add creation of directory structure per connector when creating descriptors: (`postgres.json` --> `postgres/postgres.json`)
* add configuration to add a schema filename prefix (`postgres.json` --> `debezium-postgres.json`)
* add configuration to add a schema filename suffix (`postgres.json` --> `postgres-1.9.0-SNAPSHOT.json`)
* add some minor logging (available schema formats, selected format, processed connectors) and throw errors on failed state (no schema formats found, no connectors found)
* stop calling the Schema Generator in a separate Java process when called as Maven Plugin
* remove Oracle connector schema generation as it is not annotated yet

closes https://issues.redhat.com/browse/DBZ-4665

ani-sha reviewed (approved):
LGTM @rk3rn3r! 

One question - We don't support the oracle connector in MCS yet?

gunnarmorling commented:
@ani-sha, no, not yet.

rk3rn3r commented:
@ani-sha @gunnarmorling Looks like actually Chris merged the PR last night LOL

rk3rn3r commented on debezium-schema-generator/src/main/java/io/debezium/schemagenerator/maven/SchemaGeneratorMojo.java:
```diff
@@ -48,74 +27,24 @@
-    private String classPathEntryFor(Class<?> clazz) {
-        return File.pathSeparator + clazz.getProtectionDomain().getCodeSource().getLocation().toString();
+    public void execute() {
+        new SchemaGenerator().run(format, outputDirectory.toPath(), groupDirectoryPerConnector, filenamePrefix, filenameSuffix);
```
@gunnarmorling I changed the Maven Plugin to work inlined instead of that separate JVM process.

gunnarmorling commented on debezium-schema-generator/pom.xml:
```diff
@@ -65,6 +84,41 @@
       </plugins>
     </pluginManagement>
+    <plugins>
+      <plugin>
```
Could you provide some background why the fat JAR is needed? I know you mentioned it yesterday, but I forgot :/

gunnarmorling replied:
If we truly need it, can you please add the plug-in versions to the versions / plug-in management blocks in the build parent pom rather than here?

rk3rn3r replied:
When using the plugin downstream I was not able to provide the dependencies on the classpath because of the static configured classpath before. I don't know if it resolves now?

gunnarmorling replied:
Try it out and we'll see :)

rk3rn3r replied:
@gunnarmorling is there no other way than to define all connectors plus `org.apache.kafka.connect-api` as "provided dependencies in a downstream build?

gunnarmorling replied:
@rk3rn3r, so having to define the actual connectors for downstream makes sense to me. It's the lever for controlling which files you want to generate. In regards to `connect-api`, how about you add it as a dependency to the generator plug-in? Then this should be pulled in automatically now.

gunnarmorling reviewed (commented):
Hey @rk3rn3r, some remarks inline. I've also pushed a commit for resolving the required classpath automatically (the built project's dependencies as well as the plug-in dependencies); for that I had to go back to forking the Java process though, as we cannot get the right classpath from Maven in-process. Did anything in particular speak against that forking? I think it's the better approach, as it avoids the need for specifying the current project as a dependency to the generator plug-in.

gunnarmorling commented on debezium-connector-mongodb/pom.xml:
```diff
@@ -216,6 +216,18 @@
                 <groupId>io.debezium</groupId>
                 <artifactId>debezium-schema-generator</artifactId>
                 <version>${project.version}</version>
+                <dependencies>
+                    <dependency>
+                        <groupId>${project.groupId}</groupId>
+                        <artifactId>${project.artifactId}</artifactId>
+                        <version>${project.version}</version>
+                    </dependency>
+                    <dependency>
+                        <groupId>org.apache.kafka</groupId>
+                        <artifactId>connect-api</artifactId>
+                        <version>${version.kafka}</version>
+                    </dependency>
+                </dependencies>
```
That's not needed any longer as per the commit I've pushed (same for the other modules).

rk3rn3r replied:
Removed everywhere

rk3rn3r commented on debezium-schema-generator/pom.xml:
```diff
@@ -18,6 +18,21 @@
+    <dependency>
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-core</artifactId>
+      <version>1.2.10</version>
```
Agree!

rk3rn3r commented on debezium-schema-generator/src/main/java/io/debezium/schemagenerator/SchemaGenerator.java:
```diff
@@ -34,21 +37,40 @@ public static void main(String[] args) {
     }
 
-    private void run(String formatName, Path outputDirectory) {
+    public void run(String formatName, Path outputDirectory, Boolean groupDirectoryPerConnector, String filenamePrefix, String filenameSuffix) {
```
Done

rk3rn3r commented:
@gunnarmorling I addressed your comments!

* reomved fat jar build with dependencies
* removed logback in favor of java.util.Logging

Downstream schema build works, but is there any other way than adding all connectors plus `org.apache.kafka.connect-api` as "provided" dependencies in the pom.xml?

I just realize that a change in the `schemabuilder` module don't trigger connector builds. Should a change here trigger a connector build (as integration test)?

gunnarmorling commented:
> Downstream schema build works, but is there any other way than adding all connectors plus org.apache.kafka.connect-api as "provided" dependencies in the pom.xml?

See inline. Adding connectors seems reasonable, KC API could become a plug-in dependency.

> I just realize that a change in the schemabuilder module don't trigger connector builds. Should a change here trigger a connector build (as integration test)?

Yes, that's true indeed. Logged https://issues.redhat.com/browse/DBZ-4672 for that

gunnarmorling commented on debezium-schema-generator/src/main/java/io/debezium/schemagenerator/SchemaGenerator.java:
```diff
@@ -12,43 +12,68 @@
     }
 
-    private void run(String formatName, Path outputDirectory) {
+    public void run(String formatName, Path outputDirectory, boolean groupDirectoryPerConnector, String filenamePrefix, String filenameSuffix) {
```
Can remain private now?

rk3rn3r replied:
@gunnarmorling Yes, the only case I could imagine if we want to create a Schema Generator CLI or include it in `kcctl` or another Debezium or downstream CLI, then it could make sense to access it as public. wdyt?

gunnarmorling replied:
We can change it to `public` whenever that requirement actually comes up.

gunnarmorling commented:
> reomved fat jar build with dependencies

There's still a configuration of the JAR plug-in? Should not be needed any longer?

gunnarmorling commented on debezium-schema-generator/src/main/java/io/debezium/schemagenerator/SchemaGenerator.java:
```diff
@@ -12,43 +12,68 @@
 import java.util.Optional;
 import java.util.ServiceLoader;
 import java.util.ServiceLoader.Provider;
+import java.util.logging.Logger;
```
I was thinking of `java.lang.System.Log` actually.

rk3rn3r replied:
Changed it!
I thought java.util.logging is the more powerful API falling back to System.Log? Or is it the other way around? These new APIs confuse me.
